In [33]:
import validaciones

# Ejercicios Python

## Zen of Python

Lee una vez mas el Zen of Python antes de iniciar con la tarea

In [34]:
import this

## 1 - XOR

En la siguiente celda crea una funcion para la compuerta logica XOR 

<img src="assets/xor.png" width="400">


In [35]:
def xor(a,b):
    """
    Esta funcion recibe dos valores booleanos, ejemplo xor(1, 0) o True y False
    y retorna 0 o 1 como resultado de la operacion logica XOR
    """
    # tu codigo aqui
    return int((a and not b) or (not a and b))

In [36]:
xor(0 , 1) 

1

In [37]:
xor(0 , 0)

0

In [38]:
xor(True , False)

1

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

In [39]:
validaciones.xor_function(xor)

Prueba 1
Prueba 1 [CORRECTA]
Prueba 2
Prueba 2 [CORRECTA]


## 2 - Sumando N series

Tenemos una secuencia cuyo $n^{th}$ es:
\begin{equation*}
\ T_n = n^2 - (n-1)^2
\end{equation*}

Hay que evaluar las series:
\begin{equation*}
\ S_n = T_1 + T_2 + ... + T_n
\end{equation*}


**Instrucciones**
En la siguiente celda en la funcion summingSeries debemos **retornar** en valor de $S_n$, tenemos como entrada n

**Ejemplo 1:**

Si la entrada es 2

Salida:
4

Explicacion:

$T_1 = 1$

$T_2 = 3$

$S_2 = T_1 + T_2$

$S_2 = 4 $



**Tip :**
Antes te implementar $S_n$ la solución, analicemos algebraicamente el problema para encontrar una solución equivalente optima.


In [49]:
def summatoriaSerie(n):
    """
    La funcion recibe como parametro n y retorna el valor se la serie
    """
    # tu codigo aqui  
    #list = [2*i-1 for i in range(1,n+1)]
    
    # resultado no optimo
    #list = []
    #for i in range(1,n+1):
    #    list.append(2*i-1)    
    
    #return sum(list)
    
    # la sumatoria de n es equivalente al cuadrado de n
    # ver resultado en la tabla siguiente
    
    return n**2

In [57]:
def summatoriaSerieNoOptima(n):
    """
    La funcion recibe como parametro n y retorna el valor se la serie
    """
    # tu codigo aqui  
    #list = [2*i-1 for i in range(1,n+1)]
    
    # resultado no optimo
    list = []
    for i in range(1,n+1):
        list.append(2*i-1)    
    
    return sum(list)    

In [63]:
print("T     Sumatoria      Funcion Normal")

for i in range(1,11):    
    print(str(i) + "     " +str(summatoriaSerie(i))+ "     " +str(summatoriaSerieNoOptima(i)))
    

T     Sumatoria      Funcion Normal
1     1     1
2     4     4
3     9     9
4     16     16
5     25     25
6     36     36
7     49     49
8     64     64
9     81     81
10     100     100


#### 2.1 - Ciclos

Crea un ciclo for que imprima el resultado de la funcion **summingSeries** desde 1 hasta 10

In [64]:
for i in range(1,11):
    print(summatoriaSerie(i))
    
    

1
4
9
16
25
36
49
64
81
100


### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [65]:
validaciones.summingSeries_function(summatoriaSerie)


Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]


## 3 - Factores Primos

Para cada $n$ queremos obtener el conteo maximo de numeros primos unicos en en rango inclusivo $[1, n]$,
y retornar el valor del conteo en una nueva linea

**Nota:** Recuerda que un numero primo solo es divisible por el mismo y que 1 no es un numero primo

##### Ejemplos:

###### Ejemplo 1
Entrada: 1

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,1]$ es $0$, porque $1$ no es un numero primo.

###### Ejemplo 2
Entrada: 3

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,3]$ es $1$, porque el numero $3$ tiene 1 factor primo unico (el mismo) 

###### Ejemplo 3
Entrada: 500

Salida esperada: 4

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,500]$ es $4$, porque el producto de los primeros cuatro numeros primos unicos es $2 \times 3 \times 5 \times 7 = 210$ y no hay ningun numero primo unico que multiplicado por el resultado sea $	\leqslant 500$



#### Tip: 
Utiliza la funcion `range()` de python


In [43]:
def conteoPrimos(n):
    """
    Esta funcion recibe como parametro n y 
    retorna el conteo maximo de numeros primos unicos
    """
    # tu codigo aqui
    cont = 0
    total = 1
    
    for i in range(2, n):
        if esPrimo(i):
            total = total * i
            if (total <= n):
                cont += 1
            else:
                break;
                
            
    return cont
    
def esPrimo(num):
    if(num < 2):
        return False;
    
    for i in range(2,num):
        if (num % i == 0):
            return False;
        
    return True

#print (conteoPrimos(500))

### Pruebas

In [44]:
conteoPrimos(5)

1

In [45]:
conteoPrimos(500)

4

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [46]:
validaciones.primeCount_function(conteoPrimos)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]
Prueba 7
Prueba 7[CORRECTA]
Prueba 8
Prueba 8[CORRECTA]
Prueba 9
Prueba 9[CORRECTA]
Prueba 10
Prueba 10[CORRECTA]
Prueba 11
Prueba 11[CORRECTA]
Prueba 12
Prueba 12[CORRECTA]


## 4 - Metodo de Euler

El metodo de Euler se utiliza para aproximar la solucion particular de una ecuacion diferencial dado un valor inicial. Con esta informacion se sabe que la grafica de esa solucion pasa a traves del punto $(x_0,y_0)$ y tiene una pendiente $F(x_0,y_0)$ en ese punto, esto da un punto de partida para aproximar la solucion.

A partir del punto inicial, se sigue en la direccion indicada por la pendiente. Mediante un pequeño paso $h$, se mueve a lo largo de la recta tangente hasta llegar al punto $(x_1,y_1)$ donde:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ y_1 = y_0 + hF(x_0,y_0)\\
\end{equation*}


<img src="assets/Euler.png" width="350">

Como se muestra en la figura se considera $(x_1,y_1)$ como un nuevo punto inicial.

Los valores de $x$ son:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ x_2 = x_1 + h       \\
\    .                \\
\    .                \\
\    .                \\
\ x_n = x_{n-1} + h   \\
\end{equation*}


Los valores de $y$ son:
\begin{equation*}
\ y_1 = y_0 + hF(x_0,y_0)\\
\ y_1 = y_1 + hF(x_1,y_1)\\
\       . \\
\       . \\
\       . \\
\ y_n = y_{n-1} + hF(x_{n-1},y_{n-1})\\
\end{equation*}


#### Problema 4
El ejecicio consiste en implementar la funcion *euler_metod* esta funcion recibe como parametro el 
* punto inicial $x_0$
* punto inicial $y_0$
* Paso $h$ 
* numero de iteraciones $n$ 
* La equacion diferencial que queremos aproximar, esta funcion tambien recibe los parametros $x$ y $y$
* La funcion de la solucion exacta que tambien tambien recibe los parametros $x$ y $y$

La funcion debe retornar el valor de $y_n$ y el error
$error=|y - y_n|$ 

In [6]:
import math 

In [21]:
def euler_method(x,y,h,n, diff_eq, func):
    """
    Parametros x, y punto inicial, logitud del paso h , n numero de iteraciones,
    diff_eq la equacion diferencial y func la funcion de la solucion exacta
    Retorna y y el error
    """
    # tu codigo aqui
    
    for i in range(n):
        x1 = x + h
        y1 = y + h * diff_eq(x,y)
        yOrig = func(x1)
        error = math.fabs(yOrig - y1)        
        
        print("x: " + str(x1))
        print("yn: " + str(y1))
        print("y: " + str(yOrig))
        print("error: " + str(error))
        print("============================")
        
        x = x1
        y = y1
        
    return yOrig,error
    
euler_method(0,4,0.1,10, y_prime,y)
        

x: 0.1
yn: 3.2
y: 3.2749230123119273
error: 0.0749230123119271
x: 0.2
yn: 2.56
y: 2.6812801841425573
error: 0.12128018414255726
x: 0.30000000000000004
yn: 2.048
y: 2.1952465443761056
error: 0.14724654437610551
x: 0.4
yn: 1.6384
y: 1.7973158564688863
error: 0.15891585646888617
x: 0.5
yn: 1.31072
y: 1.4715177646857693
error: 0.16079776468576923
x: 0.6
yn: 1.0485760000000002
y: 1.2047768476488085
error: 0.15620084764880837
x: 0.7
yn: 0.8388608000000002
y: 0.986387855766426
error: 0.14752705576642577
x: 0.7999999999999999
yn: 0.6710886400000001
y: 0.8075860719786218
error: 0.13649743197862163
x: 0.8999999999999999
yn: 0.5368709120000001
y: 0.6611955528863462
error: 0.1243246408863461
x: 0.9999999999999999
yn: 0.4294967296000001
y: 0.5413411329464509
error: 0.1118444033464508


(0.5413411329464509, 0.1118444033464508)

### Ejemplo

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial y}{\partial x} = -2y \\
\end{equation*}
Donde $y(0) = 4$ vamos a unsa un de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ y = 4e^{-2x} \\
\end{equation*}

In [5]:
### Implementando las funciones del ejemplo
def y_prime(x, y):
    return -2*y

def y(x):   
    return 4 * math.exp(-2*x)

In [18]:
"""
Llamamos a la funcion de euler
euler_method(0,4,0.1,10, y_prime,y)
Parametros :
x = 0
y = 4
d = 0.1
n = 10
diff_eq = y_prime
func = y
"""

euler_method(0,4,0.1,10, y_prime,y)

x:  0.1
y_1:  3.2
y:  3.2749230123119273
error:  0.0749230123119271
x:  0.2
y_2:  2.56
y:  2.6812801841425573
error:  0.12128018414255726
x:  0.3
y_3:  2.048
y:  2.195246544376106
error:  0.14724654437610596
x:  0.4
y_4:  1.6384
y:  1.7973158564688863
error:  0.15891585646888617
x:  0.5
y_5:  1.3107
y:  1.4715177646857693
error:  0.16081776468576936
x:  0.6
y_6:  1.0486
y:  1.2047768476488085
error:  0.15617684764880857
x:  0.7
y_7:  0.8389
y:  0.986387855766426
error:  0.14748785576642598
x:  0.8
y_8:  0.6711
y:  0.8075860719786215
error:  0.1364860719786215
x:  0.9
y_9:  0.5369
y:  0.6611955528863461
error:  0.12429555288634608
x:  1.0
y_10:  0.4295
y:  0.5413411329464508
error:  0.11184113294645082


### Solucion:

Utiliza esta tabla como guia para validar que tu funcion este correcta,
y reemplaza los valores que faltan con la salida de tu funcion.

| x     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| y     | 4 | 3.2749      | 2.6813 | 1.7973 | 1.2048 | 0.8076 | 0.5413       |
| $y_n$ | 4 | 3.2         | 2.56   | 1.6384 | 1.0486 | 0.6711 | 0.4294       |
| error | 0 | 0.0749      | 0.1213 | 0.1589 | 0.1562 | 0.1365 | 0.1118       |

### 4.1 Probando la implementacion con una aplicación real de data science

Podemos utilizar ecuaciones diferenciales  como modelos matemáticos de fenónemos reales, por ejemplo si llegas a trabajar como data scientist para una farmaceutica  puede que interese analizar la desintegración radioactiva de cierta sustancia y como esto cambia en el tiempo
#### Desintegracion Radiactiva

$A$ es la cantidad de sustancia en el tiempo $t$, la derivada $\frac{\partial A}{\partial t}$ es rapidez a la que se desintegran los nucleos de una sustancia.

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial A}{\partial t} =  -0.56A \\
\end{equation*}
Donde $A(0) = 1$ vamos a un tamaño de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ A(t) = e^{-0.56t} \\
\end{equation*}

**Nota** En este caso sencillo, poseemos la solución exacta a la ED por lo cual no se necesita una solución aproximada , pero en aplicaciones y proyectos reales muchas veces la solución exacta no se posee, o es muy difícil de calcular y recurrimos a soluciones aproximadas por computadora , para este ejercicio el saber la solución exacta nos sirve para medir la exactitud de nuestra solución aproximada.

In [27]:
### Implementando las funciones del ejemplo
def y_prime_new(x, y):
    return -0.56*y

def y_new(x):   
    return math.exp(-0.56*x)

In [29]:
"""
Llamamos a la funcion de euler
Parametros :
x = 0
y = 1
h = 0.1
n = 10
diff_eq = y_prime_new
func = y_new
"""

euler_method(0,1,0.1,10, y_prime_new,y_new)

x: 0.1
yn: 0.944
y: 0.9455391358903963
error: 0.001539135890396337
x: 0.2
yn: 0.8911359999999999
y: 0.8940442575003572
error: 0.002908257500357281
x: 0.30000000000000004
yn: 0.8412323839999999
y: 0.8453538346846587
error: 0.004121450684658856
x: 0.4
yn: 0.7941233704959999
y: 0.7993151343693651
error: 0.005191763873365174
x: 0.5
yn: 0.7496524617482239
y: 0.7557837414557255
error: 0.006131279707501602
x: 0.6
yn: 0.7076719238903233
y: 0.7146231058160573
error: 0.006951181925733962
x: 0.7
yn: 0.6680422961524652
y: 0.675704113960626
error: 0.007661817808160776
x: 0.7999999999999999
yn: 0.6306319275679272
y: 0.6389046840319162
error: 0.008272756463989062
x: 0.8999999999999999
yn: 0.5953165396241232
y: 0.6041093828558647
error: 0.008792843231741454
x: 0.9999999999999999
yn: 0.5619788134051723
y: 0.5712090638488149
error: 0.009230250443642563


(0.5712090638488149, 0.009230250443642563)

## 5 - Rotacion de listas hacia la izquierda

Una rotacion a la izquierda en una lista mueve cada elemento de la lista una vez hacia la izquierda. Por ejemplo, si hacemos $1$ rotacion en en el arreglo $[1,2,3,4,5]$, el arreglo resultante seria $[2,3,4,5,1]$

#### Descipcion de la funcion
Dado un arreglo $a$ de $n$ integers y un numero $d$ de rotaciones, realizar $d$ rotaciones hacia la izquierda. La funcion debe retorner el arreglo re-ordenado.

**rotIzquierda** tiene los siguientes parametros:

* $a$ Un arreglo de numeros entreros.
* $d$ numero de rotaciones.

In [27]:
def rotIzquierda(a, d):
    # tu codigo aqui
    
    for i in range(0,d):
        #print(i)
        lastItem = [a[0]]
        restItem = a[1:len(a)]
    
        a = restItem + lastItem
        #print (a)
    
    
    return a

In [28]:
rotIzquierda([1, 2, 3, 4, 5], 4)
# Resultado: [5, 1, 2, 3, 4]

[5, 1, 2, 3, 4]

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

#### TIP:
Utiliza **slice** en el array para optimizar la funcion

In [29]:
validaciones.rotIzquierda_function(rotIzquierda)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
